query can kill server

**BUG REPORT**

- What did you do?
```
rara@Andrass-MacBook-Pro:~$ docker run -d --name=cockroach -p 26257:26257 -p 9999:9999 cockroachdb/cockroach:latest  start --insecure --http-port 9999
8e7a505b775b1d10294386987b35c915717b24c8b1e03a42c4b521078d599b6d
rara@Andrass-MacBook-Pro:~$ docker exec -it cockroach bash
root@8e7a505b775b:/cockroach# grep -F '[config]' cockroach-data/logs/cockroach.log 
I170331 14:41:29.601303 1 util/log/clog.go:991  [config] file created at: 2017/03/31 14:41:29
I170331 14:41:29.601303 1 util/log/clog.go:991  [config] running on machine: 8e7a505b775b
I170331 14:41:29.601303 1 util/log/clog.go:991  [config] binary: CockroachDB CCL f9b8c0c (linux amd64, built 2017/03/31 03:01:38, go1.8)
I170331 14:41:29.601303 1 util/log/clog.go:991  [config] arguments: [/cockroach/cockroach start --insecure --http-port 9999]
./cockroach sql -e "SELECT 1 GROUP BY 1 HAVING sum(1) OVER (PARTITION BY 1) > 1;" #will kill container
Error: driver: bad connection
Failed running "sql"
root@db9e2fa13cf8:/cockroach# 
rara@Andrass-MacBook-Pro:~$
```

- What did you expect to see?
Not sure...
This is a minimal failing version of a complex query. I'm a noob at sql, so I don't know if an SQL server should support GROUP BY ... HAVING AGGREGATION(...) OVER(...) ...  queries. If not I would expect some parsing error message.
- What did you see instead?
Query above made the docker container with the sql server die, without any log messages.

Sorry, probably I've checked the other log file, and haven't found the stacktrace. So here it is:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: SELECT 1 GROUP BY 1 HAVING sum(1) OVER (PARTITION BY 1) > 1;: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: SELECT 1 GROUP BY 1 HAVING sum(1) OVER (PARTITION BY 1) > 1;: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: SELECT 1 GROUP BY 1 HAVING sum(1) OVER (PARTITION BY 1) > 1;: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xdd4892]

goroutine 362 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4203fa8c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x94
panic(0x1ae67c0, 0xc420b94a90)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc42080e000, 0x7f0368cf3c00, 0xc4204bb1a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:348 +0x74
panic(0x1ae67c0, 0xc420b94a90)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc420bec032, 0x3c)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:493 +0x131
panic(0x1b11340, 0x290bee0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/parser.(*FuncExpr).Eval(0xc4204b9440, 0xc4204f9840, 0xc4204b9440, 0x294e0c0, 0xc4204b9440, 0x30)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:2395 +0x282
github.com/cockroachdb/cockroach/pkg/sql/parser.(*ComparisonExpr).Eval(0xc420b98080, 0xc4204f9840, 0x2954500, 0xc4205246b0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:2328 +0x7a
github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).renderRow(0xc420166a20, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:481 +0x77
github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).Next(0xc420166a20, 0x7f0368cf3c00, 0xc420beeba0, 0xc420bf62f8, 0x1313528, 0x7f0368cf3c00)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:138 +0xd4
github.com/cockroachdb/cockroach/pkg/sql.(*groupNode).Next(0xc4201694a0, 0x7f0368cf3c00, 0xc420beeba0, 0x2951cc0, 0xc4201694a0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/group.go:340 +0x620
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execClassic(0xc4201201a0, 0xc4204f9800, 0x2951cc0, 0xc4201694a0, 0xc420bf64b0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1362 +0x115
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4201201a0, 0x2949d20, 0xc420beeb70, 0xc4204f9800, 0xc420beed01, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1480 +0x656
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4201201a0, 0xc420376000, 0x2949d20, 0xc420beeb70, 0x0, 0xc420000101, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1235 +0x47f
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc4201201a0, 0xc420376000, 0xc420b94770, 0x1, 0x1, 0x0, 0xc420000101, 0x0, 0x0, 0xc420bf6c40, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:944 +0x800
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4201201a0, 0xc420376000, 0xc420b94770, 0x1, 0x1, 0x0, 0x1, 0xc420b947d0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:854 +0xce
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func1(0x7f0368cf3c00, 0xc420beeba0, 0xc420522b40, 0xc420b947d0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:689 +0x1df
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc420522b40, 0x7f0368cf3c00, 0xc420beeba0, 0x10101, 0xc420524a10, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:600 +0xb6
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4201201a0, 0xc420376000, 0xc420bec032, 0x3c, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420bec000)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:711 +0x649
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4201201a0, 0xc420376000, 0xc420bec032, 0x3c, 0x0, 0x0, 0x0, 0x0, 0xc420bec000)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:499 +0x11b
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc42080e000, 0xc420bec032, 0x3c, 0x0, 0x0, 0x0, 0x0, 0xa61401, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:831 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc42080e000, 0xc42080e028, 0xc420b816e0, 0x51)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:499 +0xa9
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc42080e000, 0x7f0368cf3c00, 0xc4204bb1a0, 0xc420be6a00, 0x5400, 0xc4203b0690, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:427 +0x8b4
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4203b0540, 0x7f0368cf3ce0, 0xc4209e4840, 0x2952860, 0xc420c06000, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:415 +0x927
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x2952860, 0xc420c06000)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:587 +0x153
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4203fa8c0, 0xc42000e040, 0x2952860, 0xc420c06000, 0xc420234300)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x239

```

--- a/pkg/sql/group.go
+++ b/pkg/sql/group.go
@@ -23,7 +23,10 @@ import (
 
 	"golang.org/x/net/context"
 
+	"errors"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -98,6 +101,11 @@ func (p *planner) groupBy(
 	// Normalize and check the HAVING expression too if it exists.
 	var typedHaving parser.TypedExpr
 	if n.Having != nil {
+		if p.parser.WindowFuncInExpr(n.Having.Expr) {
+			return nil, pgerror.WithPGCode(
+				errors.New("window functions are not allowed in HAVING"),
+				pgerror.CodeWindowingError)
+		}
 		var err error
 		typedHaving, err = p.analyzeExpr(ctx, n.Having.Expr, s.sourceInfo, s.ivarHelper,
 			parser.TypeBool, true, "HAVING")

--- a/pkg/sql/testdata/logic_test/window
+++ b/pkg/sql/testdata/logic_test/window
@@ -45,6 +45,9 @@ INSERT INTO kv (k, v) VALUES (99, COUNT(1) OVER ())
 query error window functions are not allowed in WHERE
 SELECT k FROM kv WHERE AVG(k) OVER () > 1
 
+query error window functions are not allowed in HAVING
+SELECT 1 GROUP BY 1 HAVING SUM(1) OVER (PARTITION BY 1) > 1;
+
 query R
 SELECT avg(k) OVER () FROM kv ORDER BY 1
 ----

